html/webappapis/structured-clone/structured-clone-battery-of-tests.js

```diff
@@ -0,0 +1,60 @@
+/**
+ * Runs a collection of tests that determine if an API implements structured clone
+ * correctly.
+ *
+ * The `runner` parameter has the following properties:
+ * - `setup()`: An optional function run once before testing starts
+ * - `teardown()`: An option function run once after all tests are done
+ * - `preTest()`: An optional, async function run before a test
+ * - `postTest()`: An optional, async function run after a test is done
+ * - `structuredClone(obj)`: Required function that somehow structurally clones
+ *                               an object.
+ */
+async function structuredCloneBatteryOfTests(runner) {
+  const tests = [
+    {
+      description: "structuredClone returns a structural clone of an object",
+      async f(runner) {
+        const obj = {
+          a: 1,
+          b: {
+            a: 1,
+            b: [2, 3, 4]
+          },
+          c: [7, 8, 9]
+        };
+        const copy = await runner.structuredClone(obj);
+        assert_equals(JSON.stringify(obj), JSON.stringify(copy));
+        assert_true(copy !== obj);
+      }
+    },
+    {
+      description: "structuredClone handles primitives",
+      async f(runner) {
+        const copy = await runner.structuredClone(4);
+        assert_equals(copy, 4);
+      }
+    }
+  ]
+  const defaultRunner = {
+    setup() {},
+    preTest() {},
+    postTest() {},
+    teardown() {}
+  };
+  runner = Object.assign({}, defaultRunner, runner);
+
+  await runner.setup();
```

Comment:
If this `setup` function completes asynchronously, it is not guaranteed to happen before the page finishes loading. If `promise_test` doesn't run by then, I don't think the test will run correctly. (Not super experienced with WPT; let me know if there's some sneaky way this is safe.)

Instead, you could have each `promise_test` await the setup promise. (Redundantly awaiting the setup doesn't seem problematic.) 

I'd suggest `structuredCloneBatteryOfTests` being a non-async function to reduce the risk of unexpectedly awaiting before test registration in general. The only interesting waiting that happens here seems to be between waiting for all tests to finish and tearing down.

Reply:
TIL!

I like your suggestion, will implement it that way.